end_to_end_tests/golden-record/my_test_api_client/api/tests/upload_file_tests_upload_post.py

```diff
@@ -25,10 +26,12 @@ def _get_kwargs(
     files = {}
     data = {}
     for key, value in multipart_data.to_dict().items():
-        if isinstance(value, File):
+        if is_file(value):
             files[key] = value
-        else:
+        elif isinstance(value, str):
             data[key] = value
+        else:
+            data[key] = json.dumps(value)
```

Comment:
This is one question I have: should we really be JSON stringifying anything that's not a string? I don't use multipart/form so I don't know what all you can actually put in one and I haven't found the time to do the requisite reading 😬 . My thought is that anything which _can_ be put into a form `httpx` should be converting properly for us... but thats probably a naive assumption. So is there any external guidance we can follow on "how to represent x in a form"?

I also think the better move is probably to have a `to_multipart` separate from a `to_json` or something? I've been trying to avoid runtime instance checks when possible because they are slow and we already know the types statically from the OpenAPI document so we really don't need them. Obviously this will be more work but is probably the route we should take, especially if we get into the transforming-data-for-forms game.

Reply:
Good question(s)! 

Firstly, the[ OpenAPI spec](https://swagger.io/docs/specification/describing-request-body/multipart-requests/) seems to encourage having `object` fields in multipart requests, and transmitting them with `Content-Type: application/json`. I don't think `httpx` does this for us though. 

If I have a dict `foo = {"bar": "baz"}` and I call `httpx.post("my/multipart/endpoint", files=files, data=foo)` then I get `TypeError: Invalid type for value. Expected str or bytes.` which is thrown by `httpx\_multipart.py", line 24` because it expects all multipart fields to be `str` or `bytes`. Hence the stringifying - though I agree it's not the most elegant solution.

Re: `to_multipart` separate from `to_dict` (I assume that's what you meant by `to_json`) - yes it does make sense, I'll give it a go. I suppose this should take the form of a new macro in `endpoint_macros.py.jinja` similar to `json_body` and `query_params`.

Reply:
From that linked piece of the spec, it looks like there is an `encoding` property that should be set if using any special content encoding for multiple parts. Maybe the solution is to start looking at and supporting that property? I don't think we use it yet.

Reply:
Didn't spot that, yes it would probably make sense to use it eventually, but I think it's still worth having sensible defaults (i.e. `application/json` for objects and arrays). If it's ok I'd rather get this finished and merged and leave the `encoding` stuff for another day :)